### ISSUE TITLE
Flux.Zeros conflicts with Flux.loadparams!

### ISSUE THREAD

**Issue by pxl-th** [FluxML/Flux.jl#1277]
Consider a model where all convolutions have their biases disabled:

```julia
m = Chain(
    Conv((3, 3), 1 => 2, bias=Flux.Zeros()),
    Conv((3, 3), 2 => 3, bias=Flux.Zeros()),
    Conv((3, 3), 3 => 4, bias=Flux.Zeros()),
)
```

If I collect its parameters (e.g. for saving them later) and print:

```julia
mp = m |> params .|> cpu
for p in mp
    println("$(typeof(p)) $(size(p))")
end
```

in the output I will have:

```
Array{Float32,4} (3, 3, 1, 2)
Array{Bool,0} () <--- disabled bias from 1st conv
Array{Float32,4} (3, 3, 2, 3)
Array{Float32,4} (3, 3, 3, 4)
```

So parameters of the model contain disabled bias, but only from the 1st layer (since parameters are stored in `Zygote.IdSet` others are ignored).
Moreover if I want to load these parameters into the model:

```julia
Flux.loadparams!(m, mp)
```

I get error `ERROR: setindex! not defined for Flux.Zeros{Bool,0}`.

Easy fix would be to define `params!` for `Flux.Zeros` that does not add it to parameters:

```julia
params!(p::Flux.Params, x::Flux.Zeros{<:Number}, seen = Flux.IdSet()) = nothing
```

I wanted to create a PR with this fix, but this solution conflicts with [this test](https://github.com/FluxML/Flux.jl/blob/master/test/layers/conv.jl#L49) which expects `Flux.Zeros` to be in parameters.
Maybe change it to instead check that `bias` is not in gradients?

**Comment by CarloLucibello** [comment]
I don't see any reason why Zeros should be in parameters. @DhairyaLGandhi what is that test for?

**Comment by DhairyaLGandhi** [comment]
It's to make sure we don't train on bias on accident

Functor would automatically collect all the params currently, a better solution might be to teach loadparams how to handle Zeros.

**Comment by pxl-th** [comment]
Also I've found out, that if you move model to any device, then `Flux.Zeros` gets converted to actual `Array` representations and gradients w.r.t. bias are no longer `nothing`.

<details>
<summary><b>MWE to reproduce</b></summary>
<p>

```julia
device = Flux.cpu
model = Conv((3, 3), 1 => 2, bias=Flux.Zeros()) |> device
xin = randn(Float32, 4, 4, 1, 1) |> device
you = randn(Float32, 2, 2, 1, 1) |> device

model_params = Flux.params(model)
println(">>> Model parameters:")
for p in model_params
    println("$(typeof(p)) $(size(p))")
end

loss(x, y) = Flux.mse(model(x), y)
grads = Flux.gradient(model_params) do
    loss(xin, you)
end

println(">>> Gradients:")
for (k, v) in grads.grads
    println("$(typeof(k)) $(size(k)) => $(typeof(v))")
end
```
</p>
</details>

<details>
<summary><b>MWE output without moving model to any device</b></summary>
<p>

```
>>> Model parameters:
Array{Float32,4} (3, 3, 1, 2)
Flux.Zeros{Bool,0} ()
>>> Gradients:
Array{Float32,4} (3, 3, 1, 2) => Array{Float32,4}
Flux.Zeros{Bool,0} () => Nothing
```
</p>
</details>

<details>
<summary><b>MWE output when `cpu` is device</b></summary>
<p>

```
>>> Model parameters:
Array{Float32,4} (3, 3, 1, 2)
Array{Bool,0} ()
>>> Gradients:
Array{Bool,0} () => Array{Float32,0}
Array{Float32,4} (3, 3, 1, 2) => Array{Float32,4}
```
</p>
</details>

<details>
<summary><b>MWE output when `gpu` is device</b></summary>
<p>

```
>>> Model parameters:
CUDA.CuArray{Float32,4,Nothing} (3, 3, 1, 2)
CUDA.CuArray{Bool,0,Nothing} ()
>>> Gradients:
CUDA.CuArray{Float32,4,Nothing} (3, 3, 1, 2) => CUDA.CuArray{Float32,4,Nothing}
CUDA.CuArray{Bool,0,Nothing} () => CUDA.CuArray{Float32,0,CUDA.CuArray{Float32,4,CUDA.CuArray{Float32,4,Nothing}}}
```
</p>
</details>

**Comment by MacKenzieHnC** [comment]
Would solving #1294 and #1284 solve this? It seems like not separating the different "trainable" parameters is becoming a problem

**Comment by DhairyaLGandhi** [comment]
I think having a solid notion of how to
1. Load Zeros
2. Don't convert when it's moved to a device

Would be valuable. #1284 would be a great to have too

**Comment by MacKenzieHnC** [comment]
If you give someone this Julia file:
```Julia
using Flux
using BSON

model = BSON.load("arbitrary_saved_model.bson")
```

do they have enough information to reverse-engineer that model in a different language?

Maybe I'm way off-base here, but it feels like using ```Zeros``` to turn bias off is covering up for an insufficient definition somewhere

**Comment by DhairyaLGandhi** [comment]
I agree that there is missing information in the parameters. Btw if we make Zeros a mutable struct, then I think we get a lot of the same benefits with having the flexibility of not defining a Singleton `Flux.Zeros`, so it should reflect as different parameters in the params object. The fact that the bias is turned off is valuable information to maintain the reproducibility of the model, which can't be ignored, and therefore should be reflected in the params.

**Comment by DhairyaLGandhi** [comment]
> Easy fix would be to define params! for Flux.Zeros that does not add it to parameters:

If we don't add it to the parameters, how do we encode that the bias is turned off when pulling out the weights from the model? I think it would work if the entire model was saved, but `loadparams!(model, weights_with_zeros)` would fail

**Comment by MacKenzieHnC** [comment]
> If we don't add it to the parameters, how do we encode that the bias is turned off when pulling out the weights from the model?

I am kind of new to Julia. Are there no polymorphism tricks we can use? Is there any way to just define a ```Conv``` struct that doesn't have a bias property in the first place?

**Comment by MacKenzieHnC** [comment]
>Another way of looking at it is, if you see an array where you try loading it with ```nothing``` (instead of ```zeros```) you break the assumption that it might be populated with legitimate values if you train it further.

@DhairyaLGandhi what did you mean by [this comment](https://github.com/FluxML/Flux.jl/pull/1281#issuecomment-657247753) in #1281?

**Comment by DrChainsaw** [comment]
This should have been closed by #1379, I must have messed up the 'fixes' syntax.
